launch_ros/launch_ros/actions/lifecycle_node.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the LifecycleNode action."""
+
+import functools
+import logging
+from typing import cast
+from typing import List
+from typing import Optional
+from typing import Text
+
+import launch
+from launch.action import Action
+
+import lifecycle_msgs.msg
+import lifecycle_msgs.srv
+
+from .node import Node
+from ..events.lifecycle import ChangeState
+from ..events.lifecycle import StateTransition
+
+_logger = logging.getLogger(name='launch_ros')
+
+
+class LifecycleNode(Node):
+    """Action that begins executing a process and sets up event handlers for the process."""
+
+    def __init__(self, *, node_name: Text, **kwargs) -> None:
+        """
+        Construct an LifecycleNode action.
+
+        Almost all of the arguments are passed to :class:`Node` and eventually
+        to :class:`launch.actions.ExecuteProcess`, so see the documentation of
+        those classes for additional details.
+
+        This action additionally emits some event(s) in certain circumstances:
+
+        - :class:`launch.events.lifecycle.StateTransition`:
+
+            - this event is emitted when a message is published to the
+              "/<node_name>/transition_event" topic, indicating the lifecycle
+              node represented by this action changed state
+
+        This action also handels some events related to lifecycle:
+
+        - :class:`launch.events.lifecycle.ChangeState`
+
+          - this event can be targeted to a single lifecycle node, or more than
+            one, or even all lifecycle nodes, and it causes the targeted nodes
+            to change state if possible, see its documentation for more details.
+        """
+        super().__init__(node_name=node_name, **kwargs)
+        self.__rclpy_subscription = None
+        # first item is unknown state
+        self.__current_state = next(iter(ChangeState.valid_states.values()))
```

Comment:
seems like it'd be more appropriate to explicitly set `self.__current_state = ChangeState.valid_states[lifecycle_msgs.msg.State.PRIMARY_STATE_UNKNOWN]` (because we want to do this regardless of dict ordering)

Reply:
Fair enough, I think I was imagining that the "first" state might change and this would be more future proof, but that's silly (it was probably late): 6955bc8